frontend/i_types.py

```diff
@@ -0,0 +1,370 @@
+        """Initialize the tree with the text file
+
+        The tree is represented by an adjacency list,
+        where each directed edge between node x and y
+        denotes that x is a subtype of y.
+
+        The subtype relationship is reflexive and transitive.
+        """
+        self.tree = {}
+        for line in open(SubtypesTree.__file_name):
+            line = line.strip()
+            if line.startswith("#") or len(line) == 0:
+                continue
+            t1, t2 = line.split()
+            if t1 in self.tree:
+                self.tree[t1].append(t2)
+            else:
+                self.tree[t1] = [t2]
+
+    def __reachable(self, current, target, visited):
+        if current in visited or (not current in self.tree):
+            return False
+        if current == target:
+            return True
+
+        visited.add(current)
+        for adjacent in self.tree[current]:
+            if self.__reachable(adjacent, target, visited):
+                return True
+        return False
+
+    def reachable(self, source, target):
+        """Check if node2 is reachable from node1"""
+        return self.__reachable(source, target, set())
+
+
+subtypes_tree = SubtypesTree()
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # Two types are considered identical if their names exactly match.
+    def __eq__(self, other):
+        return self.get_name() == other.get_name()
+
+    def __hash__(self):
+        return hash(self.get_name())
+
+    def __repr__(self):
+        return self.get_name()
+
+class TClass(Type):
+    """Type given to a class.
+
+    Attributes:
+        name (str): Class name.
+        context (Context): Contains a mapping between variables and types within the scope of this class.
+    """
+
+    def __init__(self, name, context=Context()):
+        self.name = name
+        self.context = context
+
+    def is_subtype(self, t):
+        if not isinstance(t, TClass):
+            return False
+        return subtypes_tree.reachable(self.name, t.name)
+
+class TNumber:
+    pass
+
+class TSequence:
+    pass
+
+class TMutableSequence(TSequence):
+    pass
+
+class TImmutableSequence(TSequence):
+    pass
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TNone)
+
+    def get_name(self):
+        return "None"
+
+class TBool(TClass, TNumber):
+
+    def __init__(self):
+        self.strength = 0
+        self.name = "bool"
+
+    def get_name(self):
+        return self.name
```

Comment:
this method is the same for (almost) all subtypes of TClass, can we just have it there? the classes that need a different getName method (e.g., TList, ...) can just override it